File: ios/Sources/TofuClientCore/ApiMetaGate.swift
import Foundation

/// The GET /api/v4/meta preflight, as pure decisions. Port of ApiMetaGate.kt.
///
/// After a headless login establishes the gateway session, the app asks the
/// server which API major it requires and refuses to enter the WebView on a
/// DEFINITIVE mismatch. Fail-OPEN on partial knowledge: a 404 (a server that
/// predates the meta endpoint), a transport failure, or an unparseable body
/// never blocks — only a 200 carrying a contradicting apiMajor blocks.
/// (iOS has no min-build gate yet; the contract's minAndroidBuild applies to
/// Android only. The apiMajor check is the cross-platform one.)
public enum ApiMetaGate {

    /// Absolute meta URL under the profile's base path. Resolving the absolute
    /// META_PATH against the origin would DROP the `/proxy/<port>/` prefix of
    /// a vscode code-server deploy, so the path is appended to the base as typed.
    public static func metaUrl(_ baseUrl: String) -> String {
        var base = baseUrl
        while base.hasSuffix("/") { base.removeLast() }
        return base + "/" + ApiV4Contract.metaPath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    /// nil = compatible or unknown (never blocks on partial knowledge).
    /// Non-nil = the user-facing reason this build refuses to proceed.
    public static func incompatibilityReason(status: Int, body: String?) -> String? {
        guard status == 200, let body, !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        guard let apiMajor = intField(body, "apiMajor") else { return nil }
        if apiMajor != ApiV4Contract.apiMajor {
            return "server speaks API v\(apiMajor) but this app requires " +
                "v\(ApiV4Contract.apiMajor) — update the app"
        }
        return nil
    }

    private static func intField(_ body: String, _ field: String) -> Int? {
        Rx.group(#""\#(field)"\s*:\s*(\d+)"#, in: body).flatMap(Int.init)
    }
}

File: android/app/src/main/java/com/tofu/client/session/ApiMetaGate.kt
package com.tofu.client.session

import com.tofu.client.api.ApiV4Contract

/**
 * The GET /api/v4/meta preflight, as pure decisions.
 *
 * After a headless login establishes the gateway session, the app asks the
 * server which API major and minimum Android build it requires, and refuses
 * to enter the WebView on a DEFINITIVE mismatch — a stale app talking to a
 * newer server otherwise fails later inside the SPA as unexplained 426/404s.
 *
 * The posture is fail-OPEN on partial knowledge: a 404 (a server that
 * predates the meta endpoint), a transport failure, or an unparseable body
 * never blocks — the WebView load surfaces real problems on its own. Only a
 * 200 carrying a contradicting apiMajor, or a minAndroidBuild above this
 * build, blocks.
 *
 * org.json is not on the pure-JVM test classpath, so — exactly like
 * [TofuProbe] — fields are read with anchored regexes; the meta payload is
 * flat and backend-owned (contracts/api_v4.yaml).
 */
object ApiMetaGate {

    /**
     * Absolute meta URL under the profile's base path. Resolving the absolute
     * [ApiV4Contract.META_PATH] against the origin would DROP the
     * `/proxy/<port>/` prefix of a vscode code-server deploy, so the path is
     * appended to the base as typed instead.
     */
    fun metaUrl(baseUrl: String): String =
        baseUrl.trimEnd('/') + "/" + ApiV4Contract.META_PATH.trimStart('/')

    /**
     * null = compatible or unknown (never blocks on partial knowledge).
     * Non-null = the user-facing reason this build refuses to proceed.
     */
    fun incompatibilityReason(status: Int, body: String?, appVersionCode: Int): String? {
        if (status != 200 || body.isNullOrBlank()) return null
        val apiMajor = intField(body, "apiMajor") ?: return null
        if (apiMajor != ApiV4Contract.API_MAJOR) {
            return "server speaks API v$apiMajor but this app requires " +
                "v${ApiV4Contract.API_MAJOR} — update the app"
        }
        val minAndroid = intField(body, "minAndroidBuild") ?: return null
        // Int.MAX_VALUE = "build unknown" (pure-tier tests): never blocks.
        if (appVersionCode != Int.MAX_VALUE && minAndroid > appVersionCode) {
            return "server requires app build $minAndroid or newer " +
                "(this is $appVersionCode) — update the app"
        }
        return null
    }

    private fun intField(body: String, field: String): Int? =
        Regex("\"$field\"\\s*:\\s*(\\d+)")
            .find(body)?.groupValues?.getOrNull(1)?.toIntOrNull()
}
